clients/TypeScript/packages/client/src/TxSubmission/errors.ts

```diff
@@ -0,0 +1,478 @@
+        (item as AlreadyDelegating).alreadyDelegating !== undefined,
+      Error: class AlreadyDelegatingError extends CustomError {
+        public constructor (rawError: AlreadyDelegating) {
+          super()
+          this.message = JSON.stringify(rawError.alreadyDelegating, null, 2)
+        }
+      }
+    },
+    InsufficientFundsForMir: {
+      assert: (item: SubmitTxErrorShelley): item is InsufficientFundsForMir =>
+        (item as InsufficientFundsForMir).insufficientFundsForMir !== undefined,
+      Error: class InsufficientFundsForMirError extends CustomError {
+        public constructor (rawError: InsufficientFundsForMir) {
+          super()
+          this.message = JSON.stringify(rawError.insufficientFundsForMir, null, 2)
+        }
+      }
+    },
+    TooLateForMir: {
+      assert: (item: SubmitTxErrorShelley): item is TooLateForMir =>
+        (item as TooLateForMir).tooLateForMir !== undefined,
+      Error: class TooLateForMirError extends CustomError {
+        public constructor (rawError: TooLateForMir) {
+          super()
+          this.message = JSON.stringify(rawError.tooLateForMir, null, 2)
+        }
+      }
+    },
+    DuplicateGenesisVrf: {
+      assert: (item: SubmitTxErrorShelley): item is DuplicateGenesisVrf =>
+        (item as DuplicateGenesisVrf).duplicateGenesisVrf !== undefined,
+      Error: class DuplicateGenesisVrfError extends CustomError {
+        public constructor (rawError: DuplicateGenesisVrf) {
+          super()
+          this.message = JSON.stringify(rawError.duplicateGenesisVrf, null, 2)
+        }
+      }
+    },
+    NonGenesisVoters: {
+      assert: (item: SubmitTxErrorShelley): item is NonGenesisVoters =>
+        (item as NonGenesisVoters).nonGenesisVoters !== undefined,
+      Error: class NonGenesisVotersError extends CustomError {
+        public constructor (rawError: NonGenesisVoters) {
+          super()
+          this.message = JSON.stringify(rawError.nonGenesisVoters, null, 2)
+        }
+      }
+    },
+    UpdateWrongEpoch: {
+      assert: (item: SubmitTxErrorShelley): item is UpdateWrongEpoch =>
+        (item as UpdateWrongEpoch).updateWrongEpoch !== undefined,
+      Error: class UpdateWrongEpochError extends CustomError {
+        public constructor (rawError: UpdateWrongEpoch) {
+          super()
+          this.message = JSON.stringify(rawError.updateWrongEpoch, null, 2)
+        }
+      }
+    },
+    ProtocolVersionCannotFollow: {
+      assert: (item: SubmitTxErrorShelley): item is ProtocolVersionCannotFollow =>
+        (item as ProtocolVersionCannotFollow).protocolVersionCannotFollow !== undefined,
+      Error: class ProtocolVersionCannotFollowError extends CustomError {
+        public constructor (rawError: ProtocolVersionCannotFollow) {
+          super()
+          this.message = JSON.stringify(rawError.protocolVersionCannotFollow, null, 2)
+        }
+      }
+    }
+  }
+}
```

Comment:
Doh :facepalm: ... I was wondering why this didn't cover Allegra and Mary errors... and realized I completely forgot to add them in the JSON-schema (and in the corresponding test code, which actually only ever generate Shelley errors, so this was never caught ...). 

There's more... because Allegra and Mary are a bit bigger than Shelley, and while most constructors are the same, enough are different enough to make it quite tricky to handle in a generic manner... I'll add them A.S.A.P to the schema & test generator.

Reply:
Okay, fixed in: 773219e44689a9dbf4a72b4a58571c198e52fd69 

I've added a summary of the changes between Shelley -> Allegra and Allegra -> Mary to the CHANGELOG. It's not as bad as I remembered in the end. Generators also extended :muscle:. It'd be good to start looking at code coverage at some point, because this is something I would have probably notice before...

Reply:
> I was wondering why this didn't cover Allegra and Mary errors

Ah yeah I should have raised an issue on this.